protocols/autonatv2/src/client/handler/request.rs

```diff
@@ -0,0 +1,302 @@
+            <Self as ConnectionHandler>::ToBehaviour,
+        >,
+    >,
+    outbound: futures_bounded::FuturesSet<Result<TestEnd, Error>>,
+    queued_requests: VecDeque<DialRequest>,
+}
+
+impl Handler {
+    pub fn new() -> Self {
+        Self {
+            queued_events: VecDeque::new(),
+            outbound: FuturesSet::new(DEFAULT_TIMEOUT, DEFAULT_MAXIMUM_CAPACITY),
+            queued_requests: VecDeque::new(),
+        }
+    }
+}
+
+impl ConnectionHandler for Handler {
+    type FromBehaviour = FromBehaviour;
+
+    type ToBehaviour = ToBehaviour;
+
+    type InboundProtocol = DeniedUpgrade;
+
+    type OutboundProtocol = ReadyUpgrade<StreamProtocol>;
+
+    type InboundOpenInfo = ();
+
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(DeniedUpgrade, ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        if let Some(event) = self.queued_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        if let Poll::Ready(m) = self.outbound.poll_unpin(cx) {
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                ToBehaviour::TestCompleted(m.map_err(Error::Timeout).and_then(identity)),
+            ));
+        }
+        if !self.queued_requests.is_empty() {
+            return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(REQUEST_UPGRADE, ()),
+            });
+        }
+        Poll::Pending
+    }
+
+    fn on_behaviour_event(&mut self, event: Self::FromBehaviour) {
+        match event {
+            FromBehaviour::PerformRequest(req) => {
+                self.queued_requests.push_back(req);
+            }
+        }
+    }
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::DialUpgradeError(DialUpgradeError { error, .. }) => {
+                tracing::debug!("Dial request failed: {}", error);
+            }
+            ConnectionEvent::FullyNegotiatedOutbound(FullyNegotiatedOutbound {
+                protocol, ..
+            }) => match self.queued_requests.pop_front() {
+                Some(dial_req) => {
+                    if self
+                        .outbound
+                        .try_push(handle_substream(dial_req.clone(), protocol))
+                        .is_err()
+                    {
+                        tracing::warn!("Dial request dropped, too many requests in flight");
+                        self.queued_requests.push_front(dial_req);
```

Comment:
Don't queue it again, `FuturesSet` returns you _the old_ request. By queuing it again, you will send the same request twice.

Reply:
I solved that.